File: python/src/mpeg_o/stream_reader.py
"""``StreamReader`` — sequential reader over a run inside an .mpgo file."""
from __future__ import annotations

import h5py

from .acquisition_run import AcquisitionRun
from .spectrum import Spectrum


class StreamReader:
    """Sequential reader for a single MS run inside an ``.mpgo`` file.

    Opens the file and the named run lazily; each
    :meth:`next_spectrum` call returns a :class:`Spectrum` via the
    run's hyperslab path. Suitable for streaming through runs larger
    than memory.

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``MPGOStreamReader`` · Java:
    ``com.dtwthalion.mpgo.StreamReader``.
    """

    def __init__(self, file_path: str, run_name: str) -> None:
        self._file = h5py.File(file_path, "r")
        run_group_path = f"study/ms_runs/{run_name}"
        if run_group_path not in self._file:
            self._file.close()
            raise KeyError(f"run {run_name!r} not found in {file_path!r}")
        self._run = AcquisitionRun.open(self._file[run_group_path], name=run_name)

    @property
    def total_count(self) -> int:
        """Total number of spectra in the run."""
        return self._run.count()

    @property
    def current_position(self) -> int:
        """0-based position of the next spectrum to be read."""
        return self._run.current_position()

    def at_end(self) -> bool:
        """Return ``True`` once every spectrum has been read."""
        return not self._run.has_more()

    def next_spectrum(self) -> Spectrum:
        """Return the next spectrum and advance the cursor."""
        return self._run.next_object()

    def reset(self) -> None:
        """Reposition the cursor to 0."""
        self._run.reset()

    def close(self) -> None:
        """Close the underlying HDF5 file."""
        if self._file is not None:
            self._file.close()
            self._file = None

    def __enter__(self) -> "StreamReader":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

File: java/src/main/java/global/thalion/ttio/StreamReader.java
/* TTI-O Java Implementation / Copyright (c) 2026 The Thalion Initiative / SPDX-License-Identifier: LGPL-3.0-or-later */
package global.thalion.ttio;

import global.thalion.ttio.hdf5.Hdf5File;
import global.thalion.ttio.hdf5.Hdf5Group;
import global.thalion.ttio.providers.Hdf5Provider;

/**
 * Sequential reader for a single MS run inside an {@code .tio}
 * file. Delegates to {@link AcquisitionRun}'s Streamable methods.
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOStreamReader}, Python
 * {@code ttio.stream_reader.StreamReader}.</p>
 *
 *
 */
public final class StreamReader implements AutoCloseable {

    private Hdf5File file;
    private final AcquisitionRun run;

    /**
     * Open the {@code .tio} file at {@code filePath} and position a
     * cursor at the start of run {@code runName}.
     *
     * @param filePath absolute path to the {@code .tio} file
     * @param runName  MS-run name under {@code /study/ms_runs/}
     */
    public StreamReader(String filePath, String runName) {
        this.file = Hdf5File.openReadOnly(filePath);
        try (Hdf5Group study = file.rootGroup().openGroup("study");
             Hdf5Group runs = study.openGroup("ms_runs")) {
            // AcquisitionRun.readFrom takes StorageGroup; wrap the
            // raw Hdf5Group via the Hdf5Provider adapter.
            this.run = AcquisitionRun.readFrom(
                    Hdf5Provider.adapterForGroup(runs), runName);
        }
    }

    /** @return Total number of spectra in the opened run. */
    public int totalCount() { return run.count(); }

    /** @return Current zero-based stream cursor position. */
    public int currentPosition() { return run.currentPosition(); }

    /** @return {@code true} when no further spectra remain to be read. */
    public boolean atEnd() { return !run.hasMore(); }

    /**
     * Advance the cursor and return the next spectrum.
     *
     * @return next spectrum
     * @throws java.util.NoSuchElementException when the cursor has
     *         already reached the end of the run
     */
    public Spectrum nextSpectrum() { return run.nextObject(); }

    /** Reset the cursor to the beginning of the run. */
    public void reset() { run.reset(); }

    /** Close the underlying HDF5 file. Idempotent. */
    @Override
    public void close() {
        if (file != null) {
            file.close();
            file = null;
        }
    }
}
